tests/machine/test_node.py

```
@@ -255,3 +254,23 @@ def test_recv_socket_timeout_none(socket80):
 
 def test_recv_socket_timeout_fired(socket80):
     assert run_simulation_timeout(socket80, 50) == SimulationResult.TIMEOUT
+
+
+class FakeSocket(Socket):
+    num_close = 0
+
+    def __init__(self, port, node):
+        super().__init__(port, node)
+
+    def close(self):
+        super().close()
```

Looks like you can accomplish this effect using `Mock` or `MagicMock` and the `side_effect` attribute if you like https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect

Personally, I'm more comfortable with an explicit override like this than the method below that juggles `Socket` across namespaces, but both will work.

Also, does the code below actually depend on the real behavior of `close()`? It looks like it just counts the calls. Should there be an `assert socket.is_closed` at the end?

---

The real behaviour of `close()` removes the mapping from port number to socket instance. If `Node` erroneously holds a strong reference to this socket we are discarding, then the port is not deallocated. I think that beyond ensuring the full behaviour of `Socket.close`, what we need is to ensure that the port is truly freed once the socket instance is, in principle, destroyed.

---

Also, the use of a `(Magic)?Mock` instead of this fake will not, in itself, solve the issue raised by patching the wrong `Socket` binding. The counter-intuitive patching of `itsim.machine.node.Socket` (instead of the expected `itsim.machine.socket.Socket`) stays, or the mocking fails to happen. If you are curious, I will show you tomorrow how `unittest.mock.patch` does its work, it should shed more light on what *patching* consists in here.

---

Some experiments with variating this test suggest that not preserving the normal behaviour of `Socket.close` fails to really release the port allocated through `Node.bind`. So 8704416 provides some more light on proper resource management, but the counter-intuitive patching and fake must stay.

---

I'll approve since I'm not fussed about unusual behavior like that in unit tests, but let's take a look at it in the morning